File: model_compilers/go/regression_compiler.go
/*
Code converted from Python to Go using GPT-4o (by OpenAI)

> Original Code is written in Python!

> If you find bugs or have suggestions, please open an issue on GitHub.

Date: 10.05.2025
*/

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"os"
	"strings"
)

type SymbolicModel struct {
	BasisFunctions []string  `json:"basis_functions"`
	NFeatures      int       `json:"n_features"`
	Coefficients   []float64 `json:"coefficients"`
}

func evaluateBasisFunctions(X [][]float64, basisFunctions []string, nFeatures int) [][]float64 {
	nSamples := len(X)
	nBasis := len(basisFunctions)
	XTransformed := make([][]float64, nSamples)
	for i := range XTransformed {
		XTransformed[i] = make([]float64, nBasis)
	}

	for j, funcStr := range basisFunctions {
		for i := 0; i < nSamples; i++ {
			val := 0.0
			switch {
			case funcStr == "1":
				val = 1.0
			case strings.HasPrefix(funcStr, "log1p_x"):
				idx := parseIndex(funcStr[len("log1p_x"):])
				val = math.Log1p(math.Abs(X[i][idx]))
			case strings.HasPrefix(funcStr, "exp_x"):
				idx := parseIndex(funcStr[len("exp_x"):])
				val = math.Exp(math.Min(math.Max(X[i][idx], -10), 10))
			case strings.HasPrefix(funcStr, "sin_x"):
				idx := parseIndex(funcStr[len("sin_x"):])
				val = math.Sin(X[i][idx])
			case strings.Contains(funcStr, "^"):
				parts := strings.Split(funcStr, "^")
				idx := parseIndex(parts[0][1:])
				power := parseIndex(parts[1])
				val = math.Pow(X[i][idx], float64(power))
			case strings.Contains(funcStr, " "):
				val = 1.0
				vars := strings.Split(funcStr, " ")
				for _, v := range vars {
					idx := parseIndex(v[1:])
					val *= X[i][idx]
				}
			default:
				idx := parseIndex(funcStr[1:])
				val = X[i][idx]
			}
			XTransformed[i][j] = val
		}
	}
	return XTransformed
}

func predict(X [][]float64, model SymbolicModel) []float64 {
	XT := evaluateBasisFunctions(X, model.BasisFunctions, model.NFeatures)
	yPred := make([]float64, len(X))
	for i := range X {
		sum := 0.0
		for j, coef := range model.Coefficients {
			sum += XT[i][j] * coef
		}
		yPred[i] = sum
	}
	return yPred
}

func parseIndex(s string) int {
	var idx int
	fmt.Sscanf(s, "%d", &idx)
	return idx
}

func main() {
	data, err := os.ReadFile("outputs/california_housing_model.json")
	if err != nil {
		log.Fatal(err)
	}

	var model SymbolicModel
	if err := json.Unmarshal(data, &model); err != nil {
		log.Fatal(err)
	}

	// Generate random X for testing (10 samples)
	X := make([][]float64, 10)
	for i := range X {
		X[i] = make([]float64, model.NFeatures)
		for j := 0; j < model.NFeatures; j++ {
			X[i][j] = randFloat64()
		}
	}

	yPred := predict(X, model)
	fmt.Println("Predictions:", yPred)
}

func randFloat64() float64 {
	return float64(randInt(0, 10000)) / 10000.0
}

func randInt(min int, max int) int {
	return min + int(math.Floor(float64(max-min+1)*randNorm()))
}

func randNorm() float64 {
	return math.Abs(math.Sin(float64(os.Getpid()) + float64(os.Getppid())))
}

File: model_compilers/javascript/regression_compiler.js
/*
Code converted from Python to JavaScript using GPT-4o (by OpenAI)

> Original Code is written in Python!

> If you find bugs or have suggestions, please open an issue on GitHub.

Date: 10.05.2025
*/

function evaluateBasisFunctions(X, basisFunctions, nFeatures) {
    const nSamples = X.length;
    const XTransformed = Array.from({ length: nSamples }, () => Array(basisFunctions.length).fill(0));

    for (let i = 0; i < basisFunctions.length; i++) {
        const func = basisFunctions[i];

        for (let j = 0; j < nSamples; j++) {
            if (func === '1') {
                XTransformed[j][i] = 1;
            } else if (func.startsWith('log1p_x')) {
                const idx = parseInt(func.split('_')[1].substring(1));
                XTransformed[j][i] = Math.log1p(Math.abs(X[j][idx]));
            } else if (func.startsWith('exp_x')) {
                const idx = parseInt(func.split('_')[1].substring(1));
                const val = Math.max(-10, Math.min(10, X[j][idx]));
                XTransformed[j][i] = Math.exp(val);
            } else if (func.startsWith('sin_x')) {
                const idx = parseInt(func.split('_')[1].substring(1));
                XTransformed[j][i] = Math.sin(X[j][idx]);
            } else if (func.includes('^')) {
                const [varName, power] = func.split('^');
                const idx = parseInt(varName.substring(1));
                XTransformed[j][i] = Math.pow(X[j][idx], parseInt(power));
            } else if (func.includes(' ')) {
                const vars = func.split(' ');
                let result = 1;
                for (const v of vars) {
                    const idx = parseInt(v.substring(1));
                    result *= X[j][idx];
                }
                XTransformed[j][i] = result;
            } else {
                const idx = parseInt(func.substring(1));
                XTransformed[j][i] = X[j][idx];
            }
        }
    }

    return XTransformed;
}

function predict(X, symbolicModel) {
    const XTransformed = evaluateBasisFunctions(X, symbolicModel.basis_functions, symbolicModel.n_features);
    const yPred = XTransformed.map(row =>
        row.reduce((sum, val, idx) => sum + val * symbolicModel.coefficients[idx], 0)
    );
    return yPred;
}

// Example usage:
fetch('outputs/california_housing_model.json')
    .then(response => response.json())
    .then(symbolicModel => {
        const nSamples = 10;
        const nFeatures = symbolicModel.n_features;
        const X = Array.from({ length: nSamples }, () =>
            Array.from({ length: nFeatures }, () => Math.random())
        );
        const yPred = predict(X, symbolicModel);
        console.log(yPred);
    });
